### ISSUE TITLE
After selecting an item in the picker, the 'Unfocused' event does not trigger.

### ISSUE THREAD

**Issue by bhuwancb99** [dotnet/maui#21704]
### Description

When selecting any item inside the picker, the 'Unfocused' event fails to trigger. Previously, in Xamarin Forms, it worked correctly. However, currently, it functions properly only on iOS. On Android, the event does not occur as expected; upon selecting an item, no action is taken.
![image](https://github.com/dotnet/maui/assets/65246844/e385b501-05a6-47d8-9e40-7ba92a7db87e)

![image](https://github.com/dotnet/maui/assets/65246844/381365d5-8023-4da2-9aa6-cf631fc6b85f)

![image](https://github.com/dotnet/maui/assets/65246844/8ebc2b93-e1ab-40e2-a76e-fbfa311c4ed8)

### Steps to Reproduce

Create .NET MAUI New Project -> Select .NET 8.0 -> Add Picker Inside Page - > add Event Unfocused -> When select any item inside picker -> Unfocused event not trigger.  

### Link to public reproduction project repository

https://1drv.ms/f/s!Alibu8GDjP9aggKpS2vWOW7lZMxh?e=zt5toB

### Version with bug

8.0.14 SR3.1

### Is this a regression from previous behavior?

Yes, this used to work in Xamarin.Forms

### Last version that worked well

Unknown/Other

### Affected platforms

Android

### Affected platform versions

Android 14

### Relevant log output

sometime getting this error when select picker-
Android.Views.InflateException: 'Binary XML file line #23 in com.abc.xyz:layout/mtrl_alert_select_dialog_item: Binary XML file line #23 in com.abc.xyz:layout/mtrl_alert_select_dialog_item: Error inflating class TextView'

**Comment by jaosnz-rep** [comment]
Can repro this issue at Android platform on the latest 17.10 preview 2(8.0.14).

**Comment by bhuwancb99** [comment]
facing same issue on the latest 17.10 preview 2(8.0.14). not able to trigger 'Unfocused' event after selected items inside picker.

**Comment by bhuwancb99** [comment]
Could you please provide an update on this? My application release is pending due to this change. If you have any suggestions for patching this issue, I urgently need a solution. Thank you.

**Comment by bhuwancb99** [comment]
any update?

**Comment by bhuwancb99** [comment]
any update?

**Comment by bhuwancb99** [comment]
Hi @jaosnz-rep any update on this ? I'm waiting for this fix before releasing the application to the Play Store.

**Comment by bhuwancb99** [comment]
Hi @jaosnz-rep  -any update on this bug?

**Comment by jsuarezruiz** [comment]
As a workaround, could use the SelectedIndexChanged event and use the Unfocus method.

**Comment by bhuwancb99** [comment]
Hi @jsuarezruiz  I can't use the 'SelectedIndexChanged' event because I am changing the SelectedIndex based on the entry value, so both the picker and entry's Unfocused events are not working. It's difficult for me to manage this across the entire application. After upgrading from Xamarin to MAUI, it's not possible to change the logic in many places. If possible, could you suggest any patch or code that would make the Unfocused event work?

**Comment by bhuwancb99** [comment]
Hi @jaosnz-rep -any update on this bug?

**Comment by bhuwancb99** [comment]
Hi @jaosnz-rep  and @Eilon any update on this. i am waiting to release application inside playstore.

**Comment by bhuwancb99** [comment]
Hi @jaosnz-rep and @Eilon any update on this. i am waiting to release application inside playstore.

**Comment by bhuwancb99** [comment]
any update on this when picker unfocused issue fixed?

**Comment by bhuwancb99** [comment]
It's been almost 3-4 weeks since this issue was reported. This is my priority issue because I am not releasing my upgraded app from Xamarin Forms to MAUI due to this pending issue. Please provide a solution for this, or suggest any patch or alternate way to fix this issue and release the app on the Play Store.

**Comment by Eilon** [comment]
@Redth / @PureWeen - any thoughts on how this issue can make progress? Would we accept a PR to this? Or know of a better workaround?

**Comment by bhuwancb99** [comment]
Hi @Eilon / @Redth / @PureWeen - any update on this ?

**Comment by bzd3y** [comment]
@bhuwancb99 Wait, why would selecting something trigger `Unfocused`? The control still has focus when something was selected, right? Why would it lose focus or report that it lost focus?

Xamarin might have done this, but that seems to be the wrong behavior (unless I'm misunderstanding something).

If you want to reproduce this for legacy reasons, I think you should be able to do it pretty easily with a MAUI handler.

**Comment by bhuwancb99** [comment]
@bzd3y 
inside a collection view, I am using a Picker to change values like item quantity or other values. When users add multiple items (like 10 to 15) to the cart, the SelectedIndexChanged event is triggered every time the page loads or when the collection view refreshes, causing 10-15 API calls to be made. However, the Unfocused event for the Picker is only called when a specific Picker value is changed, and it is not triggered when the collection view refreshes or loads for the first time.

I want to use the Picker's Unfocused event (or an alternative method) to achieve this behavior without triggering multiple API calls during page load or collection view refresh. Additionally, the Unfocused event for the Entry textbox is also not being called.

Please suggest an alternative way to handle this scenario or a method to achieve this using handlers.

**Comment by uhbtl** [comment]
I am facing the exact same problem as @bhuwancb99 since I upgraded my app from Xamarin to MAUI. I used the Unfocused event to check if the user changed the selected value too. I cannot release my MAUI app this way.

**Comment by bhuwancb99** [comment]
In Xamarin.Forms, the Unfocused event is working, and based on this, we delivered the application to the client. However, in .NET MAUI, the Unfocused event does not trigger. This is why I cannot deliver this app to the Play Store. If you are providing this functionality in Xamarin.Forms, why is it not available in MAUI? I opened this issue in April, and it has been almost 4-5 months without a solution. Three of my projects are pending release to production due to this issue, and my clients need the Unfocused functionality.

Please provide a solution for this or suggest a patch or any alternative way.

**Comment by bhuwancb99** [comment]
Yes, @uhbtl I am facing the exact same issue. If the issue is fixed, I will release all three of my applications on the Play Store or Apple Store.

**Comment by Eilon** [comment]
Hi, I think that logically the Unfocused event is not the right one to use. That event is meant for when the user unfocuses a control and focuses another control. I think you're actually looking for SelectedIndexChanged, but your issue is that this event happens even when the user didn't actually pick something? Is that correct?

If this is true, could you create a control derived from Picker that has an additional event that is only raised when the actual item changes? Or you could handle this within your app's ViewModel where even though there is an event that is raised too often, the ViewModel can compare its internal data to the UI, and detect if something *actually* changed?

**Comment by bhuwancb99** [comment]
Hi @Eilon 
I don't know why you're not providing the same Xamarin. Forms functionality in .NET MAUI. It's very difficult to manage because, in the Unfocused event, when an item is selected, an API call is made. So, please provide the same Xamarin. Forms functionality in MAUI. Also, the Unfocused event was working in .NET MAUI 7.0, but it's not working in .NET 8.0. Why is this happening? and why you are not fixing this issue on .NET 8.0

**Comment by uhbtl** [comment]
> Hi, I think that logically the Unfocused event is not the right one to use. That event is meant for when the user unfocuses a control and focuses another control. I think you're actually looking for SelectedIndexChanged, but your issue is that this event happens even when the user didn't actually pick something? Is that correct?

That´s correct.

> If this is true, could you create a control derived from Picker that has an additional event that is only raised when the actual item changes?

This is exactly what I did in Xamarin.Forms. After the user changed the selected value manually, the Unfocused Event was raised. By raising this event, the IsFocused property changed to false. So you could handle that in the OnPropertyChanged event of the picker itself (and fire your custom event).
But in MAUI, after the user changed the selected value manually, no property is changed. I found no way to handle that.

**Comment by LeoJHarris** [comment]
Facing the same issue.

**Comment by SleepyDevelops** [comment]
When looking into this issue I noticed that the IsFocused is always false, even if the Picker is open. That's probably why the unfocused is not triggered. 
Which makes this even worse i think.

**Comment by bhuwancb99** [comment]
Hi @Eilon / @Redth / @PureWeen / @bzd3y, any update on this ?

**Comment by bzd3y** [comment]
@bhuwancb99 Sorry I am so late on this. I'm also waiting on this, but it isn't quite as critical at the moment for me and I moved on to other stuff.

I think I understand what you are trying to do now. I think what you are trying to do could work without `Unfocused` getting called on selection. You'd just see the change when they moved focus to something else and were "done" with the `Picker`.

So are you saying that doesn't work?

As for some possible alternatives, one thing you might be able to do is not bind to `SelectedIndexChanged` initially, like in XAML, but do it manually in the backend C# after everything else is set up (I assume the list is initially populated) so that way it doesn't trigger for every item in the list when the `Picker` is set. And so once the list is populated, you would go and subscribe to that event for each `Picker` in the list. But depending on how things are set up, that might be tricky/tedious.

Another solution (probably the ideal solution) would be to use a handler. You mentioned it being hard to change this everywhere, but if you could add a pretty simple custom handler for `Picker`s or add mapper to the default `PickerHandler` that probably does it. It might be as simple as mapping `SelectedIndexChanged` and calling `Unfocus()` on the `Picker`.

Sorry again to take so long getting back to this.

**Comment by LeoJHarris** [comment]
I had the same issue with the TimePicker component on android but using the following solved my issue.

```csharp
public partial class TimePickerWithFocusHandler : TimePickerHandler
{
    private TimePickerDialog? _dialog;

    protected override TimePickerDialog CreateTimePickerDialog(int hour, int minute)
    {
        _dialog = base.CreateTimePickerDialog(hour, minute);

        _dialog.ShowEvent += onDialogShown;
        _dialog.DismissEvent += onDialogDismissed;

        return _dialog;
    }

    protected override void DisconnectHandler(MauiTimePicker platformView)
    {
        base.DisconnectHandler(platformView);
        _dialog = null;
    }

    private void onDialogShown(object? sender, System.EventArgs e)
    {
        VirtualView.IsFocused = true;

        if (_dialog is not null)
        {
            _dialog.ShowEvent -= onDialogShown;
        }
    }

    private void onDialogDismissed(object? sender, System.EventArgs e)
    {
        VirtualView.IsFocused = false;

        if (_dialog is not null)
        {
            _dialog.DismissEvent -= onDialogDismissed;
        }
    }
}
```

**Comment by bhuwancb99** [comment]
Hi @LeoJHarris, could you please share some code for the above solution to provide a better understanding for me or others?

**Comment by bhuwancb99** [comment]
@bzd3y - I’ve tried multiple times, but it’s still not working. I also tried different NuGet packages, but none of them worked. One package did trigger the Unfocused and Focus events, but the SelectedItem is not getting updated—it always returns the previous selected item. Here's the link to the package: https://enisn-projects.io/docs/en/uranium/latest/themes/material/components/PickerField

I’ve reported this issue regarding the SelectedItem not updating: https://github.com/enisn/UraniumUI/issues/646

My four apps, which have been migrated from Xamarin to MAUI, are pending release on the Play Store due to this issue. Right now, Xamarin supports Android 14, so we can release for that, but if Android 15 is released, it will be difficult to manage both the apps and the clients.

**Comment by bzd3y** [comment]
@LeoHarris Thanks for sharing that code. You may be able to make that shorter or not need a separate class by using the default `TimePickerHandler` and adding a mapping, but that is just a suggestion. Sometimes it is definitely easier to just use a custom handler.

@bhuwancb99 I might not be understanding what is going on then. I was going off of your first sentence in this thread:

> When selecting any item inside the picker, the 'Unfocused' event fails to trigger.

My understanding was that in Xamarin you were using this to avoid your event being fired when `SelectedIndex` changed for the first time for all of the `Pickers` on the page.

I think you can accomplish that with something like the following code, called from where your MauiAppBuilder is used. It doesn't have to be called directly from there, but maybe should be in a platform specific class or at least `#if ANDROID`.

```csharp
.ConfigureMauiHandlers(handlers =>
{
	PickerHandler.Mapper.AppendToMapping(nameof(Picker.SelectedIndex), (h, p) =>
	{
		if (p.IsFocused) // it makes sense to check this anyway since Unfocused() would be unnecessary, but this also may save you from the multiple initial calls when SelectedIndex changes for the first time.
		{
			p.Unfocus();
		}
	});
})
```

After adding that, all of the `Picker`s in your app will call `Unfocus()` when the selected index changes and (unless I am missing something) your old code that used `Unfocused` should just work.

I have tested that this works as far as being called when the `SelectedIndex` changes and that it unfocuses the `Picker`, but I haven't tested that it then calls the `Unfocused` event in such a way that you can then access the new `SelectedIndex` but I don't see why that wouldn't work.

**Comment by bhuwancb99** [comment]
@bzd3y, as per the suggested code above, the Unfocused event fires, but the **SelectedItem** remains the same. When the Unfocused event fires, I want to get the updated **SelectedIndex**. Is there any chance to get the updated **SelectedItem** as well?

**Comment by bzd3y** [comment]
Then I think something else is going on. The `SelectedItem` should get changed when the `SelectedIndex` changes, relative to the `Unfocused` getting called, anyway.

I just did some further testing and I am not seeing this. I see the new value.

A couple ideas:

This may have to do with how the `Picker.Items` are populated. Looking at code, there is some slightly different behavior depending on whether `ItemSource` is used or the items are added directly to the `Items` property (the testing I just did used this method). I don't see anything that stands out that would cause what you are seeing though.

Another idea might just be in how things are bound. Is the binding in XAML? There are some conversions that can get involved there. How is your `Unfocused` event set? Is that in XAML or in C#?

I was adding it in the C# and I see the new value in the event.

Further, if it is in C# code, is it an anonymous function where there might be some capturing going on that captures the "old" value when an anonymous function was added to the event?

**Comment by bhuwancb99** [comment]
@bzd3y - Attached is the sample project for your reference. When an item is selected inside the picker, the Unfocused event is triggered. However, the first selected item is null. When selecting another item, the picker displays the previously selected item.

[PickerUnfocusedIssueApp.zip](https://github.com/user-attachments/files/17055991/PickerUnfocusedIssueApp.zip)

**Comment by bzd3y** [comment]
Is there a chance this has to do with your binding to the CountryViewModel? This is only after looking briefly, for the same of time for both of us. When I get more time I can try to look more.

But you have this:

```xaml
<ContentPage.BindingContext>
        <vm:CountryViewModel />
    </ContentPage.BindingContext>

    <StackLayout Padding="10">
        <Label Text="Select a Country:" FontSize="Medium" />
        <Picker ItemsSource="{Binding Countries}"
                ItemDisplayBinding="{Binding Name}"
                SelectedItem="{Binding SelectedCountry}" Unfocused="Picker_Unfocused" />
        <Label Text="{Binding SelectedCountry.Name}" FontSize="Large" TextColor="Blue" />
    </StackLayout>
```

But in the C# code behind that page you have:

```csharp
    private void Picker_Unfocused(object sender, FocusEventArgs e)
    {
        var picker = sender as Picker;
        if (picker != null)
        {
            var selectedCountry = picker.SelectedItem as CountryModel;
            if (selectedCountry != null)
            {
                DisplayAlert("Selected Country", $"You selected: {selectedCountry.Name}", "OK");
            }
        }
    }
```

So you are binding the `SelectedItem` to `CountryViewModel.SelectedCountry` but then in the `Unfocused` event you are checking `SelectedItem` directly. It seems like there could be a chance something is going on there.

Also, this is using `ItemSource`, which I hadn't tested with. So when I get the time I'll try testing the code I provided with that case.

**Comment by bhuwancb99** [comment]
@bzd3y, I am waiting for your response, but in the meantime, I am using SelectedIndex or binding SelectedItem to save the response data.

**Comment by bzd3y** [comment]
@bhuwancb99 I tried your code and I do see what you are seeing. When I get some time I'll try to figure out what is going on there.

**Comment by bzd3y** [comment]
@bhuwancb99 Try this instead:

```csharp
	PickerHandler.Mapper.AppendToMapping("UnfocusedFix", (h, p) =>
	{
		if (p is Picker picker)
		{
			picker.SelectedIndexChanged += (s, e) =>
			{
				if (picker.IsFocused)
				{
					picker.Unfocus();
				}
			};
		}
	});
```

The key there is to not alter the mapping to `SelectedIndexChanged` and instead just add a custom mapper that adds a delegate to the event that will call `Unfocused()` on `SelectedIndexChanged` if the `Picker` has focus.

With that being said, the `SelectedItem` not being set by the time the Mapper calls `SelectedIndexChanged` seems like the actual bug here. You may want to create another issue for it. I tried looking into why it wasn't getting set, but I haven't figured that out yet.

**Comment by bhuwancb99** [comment]
@bzd3y - Thank you! Now it’s working as expected everywhere, and all my queries have been resolved. I truly appreciate your help with this solution, as I’ve been waiting for the last four months to release the app in the Play Store. Thank you so much again!

**Comment by bzd3y** [comment]
@bhuwancb99 You're welcome. Did you or are you going to create a bug for the `SelectedItem` issue? If not, then I probably will.

**Comment by bhuwancb99** [comment]
@bzd3y, sure you can create a bug for `SelectedItem` and link it with this issue thread.

**Comment by letscodewithkalyan** [comment]
Any update on this, I to facing similar issue..
When we add handler for SelectedItem or SelectedIndexChanged
When we pick same item above events will not fire. In that case Unfocused will not work

**Comment by bhuwancb99** [comment]
@letscodewithkalyan - Check the last 4 comments and add this code inside the MauiProgram.cs file. Then it will work as expected.

`PickerHandler.Mapper.AppendToMapping("UnfocusedFix", (h, p) =>
	{
		if (p is Picker picker)
		{
			picker.SelectedIndexChanged += (s, e) =>
			{
				if (picker.IsFocused)
				{
					picker.Unfocus();
				}
			};
		}
	});`

**Comment by bhuwancb99** [comment]
@bzd3y - The DatePicker Unfocused event is also not working. I tried using the same Unfocused event code for the DatePicker as I did for the Picker, but it isn't triggering. Do you have any solution for this?

**Comment by bzd3y** [comment]
@bhuwancb99 I assume you changed the type check to look for `DatePicker`s?
If so, and if I recall correctly, things like the `DatePicker` and `TimePicker` work a little differently. They lose focus on their own when they display the popup to allow the user to pick a value. I would look in the `PlatformView` at the "extra" events to see if you can use something there. So something like this:

```chsarp
			DatePickerHandler.Mapper.AppendToMapping("UnfocusedFix", (h, p) =>
			{
				if (p is DatePicker picker)
				{
					//h.PlatformView.<try one of the events Intellisense gives you>
				}
			});
```

When I get a chance I'll try looking for myself to see if I can find something that works.

**Comment by bhuwancb99** [comment]
@bzd3y - as per suggested but it's not working.
